### PR TITLE
fix: registration should accept hydra login

### PR DESCRIPTION
**Pull request by Benehiko**
## Related issue(s)

https://github.com/ory-corp/cloud/issues/5493

Using `propt=registration` on an OAuth2 Hydra login flow, the request is routed to the Kratos registration handler. This handler immediately redirects to the logout flow when there is an already existing session, which fails and ends up at the base `/welcome` page on the Account Experience (AX). 

This PR adds the functionality of accepting the login request in the registration handler if the user has a session and if the hydra `skip=true` is in the login request. 

When the `skip=false` is set, the request is redirected to the login flow with the login_challenge, which will complete the flow normally.
 

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: selfservice/flow/registration/handler.go [hunk 1]
```diff
@@ -318,29 +318,79 @@ type createBrowserRegistrationFlow struct {
...
+			r.URL.RawQuery = q.Encode()
+		}
+	}
+
+	if sess, err := h.d.SessionManager().FetchFromRequest(ctx, r); err == nil {
+
+		if hydraLoginRequest != nil {
+			// hydra indicates that we cannot skip the login request
+			// so we must perform the login flow.
+			if !hydraLoginRequest.GetSkip() {
```
Comment (hperl) [inline selfservice/flow/registration/handler.go:366]: I think skip is used purely for the consent. For forcing the login, we need to look for `prompt=login`. Prompt can have space-delimited values, so `prompt=login registration` would be valid.

File: selfservice/flow/registration/handler.go [hunk 2]
```diff
@@ -318,29 +318,79 @@ type createBrowserRegistrationFlow struct {
...
+			if err != nil {
+				h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, err)
+				return
+			}
+			returnTo, err := url.Parse(rt)
+			if err != nil {
+				h.d.SelfServiceErrorManager().Forward(r.Context(), w, r, errors.WithStack(herodot.ErrInternalServerError.WithReasonf("Unable to parse URL: %s", rt)))
+				return
+			}
+			x.AcceptToRedirectOrJSON(w, r, h.d.Writer(), err, returnTo.String())
```
Comment (hperl) [inline selfservice/flow/registration/handler.go:393]: I don't think we ever need to return JSON here.

**Review by hperl — CHANGES_REQUESTED** [review]
Thanks for looking into this! The approach is sensible, I just think we should look for `prompt=login` instead of `skip`.

File: selfservice/flow/registration/handler.go (hunk 1 above)
Comment (Benehiko) [reply selfservice/flow/registration/handler.go:366]: what should we then do if `skip=false`? In this case the user cannot go back to the client, on the Login flow we usually do `refresh=true` here. So this is why I redirect to login and have the login logic handle what should happen to the OAuth flow. 

In this case, the user is re-prompted to login or "confirm it's you".

File: selfservice/flow/registration/handler.go (hunk 2 above)
Comment (Benehiko) [reply selfservice/flow/registration/handler.go:393]: We do, in the case of an SPA we return 422 with a browser redirect URL - just like in the Login flow.

Comment (Benehiko) [reply selfservice/flow/registration/handler.go:393]: https://github.com/ory/kratos/blob/master/selfservice/flow/login/handler.go#L479-L496

**Review by hperl — APPROVED** [review]
LGTM!

**Pull request merged**